wasm/jit/jit_amd64.go

```diff
@@ -1182,6 +1169,122 @@ func (b *amd64Builder) handleLoad(o *wazeroir.OperationLoad) error {
+	// Note that Load8 is only for integer types.
+	moveFromMemory := b.newProg()
+	moveFromMemory.As = x86.AMOVB // 1 byte!
+	moveFromMemory.To.Type = obj.TYPE_REG
+	moveFromMemory.To.Reg = reg
+	moveFromMemory.From.Type = obj.TYPE_MEM
+	moveFromMemory.From.Reg = reservedRegisterForMemory
+	moveFromMemory.From.Index = reg
+	moveFromMemory.From.Scale = 1
+	b.addInstruction(moveFromMemory)
+	top := b.locationStack.pushValueOnRegister(reg)
+
+	switch o.Type {
+	case wazeroir.SignedInt32, wazeroir.SignedUint32:
+		top.setValueType(wazeroir.UnsignedTypeI32)
+	case wazeroir.SignedInt64, wazeroir.SignedUint64:
+		top.setValueType(wazeroir.UnsignedTypeI64)
+	}
+	return nil
+}
+
+func (b *amd64Builder) handleLoad16(o *wazeroir.OperationLoad16) error {
+	base := b.locationStack.pop()
+	if err := b.ensureLocatedOnGeneralPurposeRegister(base); err != nil {
+		return err
+	}
+
+	// At this point, base's value is on the integer general purpose reg.
+	// We reuse the register below, so we alias it here for readability.
+	reg := base.register
+
+	// We have to calculate the offset on the memory region.
+	addOffsetToBase := b.newProg()
+	addOffsetToBase.As = x86.AADDL // 32-bit!
+	addOffsetToBase.To.Type = obj.TYPE_REG
+	addOffsetToBase.To.Reg = reg
+	addOffsetToBase.From.Type = obj.TYPE_CONST
+	addOffsetToBase.From.Offset = int64(o.Arg.Offest)
+	b.addInstruction(addOffsetToBase)
+
+	// Then move a byte at the offset to the register.
+	// Note that Load16 is only for integer types.
+	moveFromMemory := b.newProg()
+	moveFromMemory.As = x86.AMOVW // 2 byte!
+	moveFromMemory.To.Type = obj.TYPE_REG
+	moveFromMemory.To.Reg = reg
+	moveFromMemory.From.Type = obj.TYPE_MEM
+	moveFromMemory.From.Reg = reservedRegisterForMemory
+	moveFromMemory.From.Index = reg
+	moveFromMemory.From.Scale = 1
+	b.addInstruction(moveFromMemory)
+	top := b.locationStack.pushValueOnRegister(reg)
+
+	switch o.Type {
+	case wazeroir.SignedInt32, wazeroir.SignedUint32:
+		top.setValueType(wazeroir.UnsignedTypeI32)
+	case wazeroir.SignedInt64, wazeroir.SignedUint64:
+		top.setValueType(wazeroir.UnsignedTypeI64)
+	}
+	return nil
+}
+
+func (b *amd64Builder) handleLoad32(o *wazeroir.OperationLoad32) error {
+	base := b.locationStack.pop()
+	if err := b.ensureLocatedOnGeneralPurposeRegister(base); err != nil {
+		return err
+	}
+
+	// At this point, base's value is on the integer general purpose reg.
+	// We reuse the register below, so we alias it here for readability.
+	reg := base.register
+
+	// We have to calculate the offset on the memory region.
+	addOffsetToBase := b.newProg()
+	addOffsetToBase.As = x86.AADDL // 32-bit!
+	addOffsetToBase.To.Type = obj.TYPE_REG
+	addOffsetToBase.To.Reg = reg
+	addOffsetToBase.From.Type = obj.TYPE_CONST
+	addOffsetToBase.From.Offset = int64(o.Arg.Offest)
+	b.addInstruction(addOffsetToBase)
+
+	// Then move a byte at the offset to the register.
+	moveFromMemory := b.newProg()
+	moveFromMemory.As = x86.AMOVL // 3 byte!
```

Comment:
I thought I understood comments like 1 byte, 2 byte (as load8, load16), but why 3 byte? guess it means I didn't understood the earlier ones!

Reply:
load32 means reading 32-bit (= 3 bytes) from the offset!